### PR TITLE
fix: raise syntax error when function name doesn't match closing function name

### PR DESCRIPTION
**Pull request by gxyd**
The title itself

### CODE REVIEW

File: src/lfortran/parser/parser.yy [hunk 1]
```diff
@@ -975,62 +976,62 @@ function
...
+                SPLIT_STMT(p.m_a, $10), $11, $12, @$); }
     | KW_FUNCTION id "(" id_list_opt ")"
         bind
         result_opt
         sep use_statement_star import_statement_star implicit_statement_star decl_statements
         contains_block_opt
         end_function sep {
-            LLOC(@$, @15); $$ = FUNCTION0($2, $4, $7, $6, TRIVIA($8, $15, @$),
-                $9, $10, $11, SPLIT_DECL(p.m_a, $12), SPLIT_STMT(p.m_a, $12), $13, @$); }
+            LLOC(@$, @14); $$ = FUNCTION0($2, $4, $7, $6, TRIVIA($8, $15, @$),
```
Comment (gxyd) [inline src/lfortran/parser/parser.yy:988]: I made this change from '15' -> '14' for correcting the position for "end_function".

Now the below program:
```f90
program expr2
    implicit none
    real :: x
    x = 1.2
    print *, helloWorld(5.1, 2)
    !call my_func(5)

contains
    subroutine my_func()
        print *, "hi"
    end subroutine

    function helloWorld(x)
        real, intent(in) :: x
        helloWorld = x + 1
    end function helloWorld

end program expr2
```

on current `main` the error is:
```bash
semantic error: Return type not specified
  --> incorrect_number_args.f90:13:5 - 17:1
   |
13 |        function helloWorld(x)
   |        ^^^^^^^^^^^^^^^^^^^^^^...
...
   |
17 |    
   | ...^ 


```

on this branch the error is:
```bash
semantic error: Return type not specified
  --> incorrect_number_args.f90:13:5 - 16:27
   |
13 |        function helloWorld(x)
   |        ^^^^^^^^^^^^^^^^^^^^^^...
...
   |
16 |        end function helloWorld
   | ...^^^^^^^^^^^^^^^^^^^^^^^^^^^ 


```

Comment (Thirumalai-Shaktivel) [reply src/lfortran/parser/parser.yy:988]: Yes, that's the way to fix it

**Comment by Thirumalai-Shaktivel** [comment]
The failure is in test_parse.cpp:1307, just do
```
-     CHECK(result->loc.last == 54);
+    CHECK(result->loc.last == 53); 
```

**Comment by gxyd** [comment]
> The failure is in test_parse.cpp:1307, just do

Ah ok, got it. Yesterday, it looked like a mysterious fail to me (couldn't figure it out, I think I was tired). Thanks @Thirumalai-Shaktivel

**Comment by Thirumalai-Shaktivel** [comment]
Cool, go ahead update the branch and resolve the git conflict.

**Review by Thirumalai-Shaktivel — APPROVED** [review]
LGTM, thank you!

**Comment by certik** [comment]
Thanks! I think this fix is good. I think it will most likely slow down valid code a little bit, but hopefully by not too much. Once we reach beta, we can revisit this. There are two competing requirements: fast compilation of valid code, and reporting all errors. For now let's report all errors, at the expense of speed of compilation of valid (error-free) code.

**Comment by gxyd** [comment]
Do we have a provision currently for a benchmarking of LFortran so that there is no significant decrease in speed from a particular patch?

I came across [this blogpost LFortran Compiles dftatom
](https://lfortran.org/blog/2023/10/lfortran-compiles-dftatom/), is the idea to use these packages of Fortran (apart from "being able to compile them") to use them for benchmarking purposes as well?


> For now let's report all errors, at the expense of speed of compilation of valid (error-free) code.

makes sense

**Comment by gxyd** [comment]
@Thirumalai-Shaktivel , just a gentle request: if you can always squash my commits. I usually (as in this case) prefer single commit PR's.

**Comment by Thirumalai-Shaktivel** [comment]
This PR had three different changes and the commit history was clean, so I merged it as is.
It would very helpful for the developers later, while they look into the history for why the changes were made.

**Comment by certik** [comment]
Yes, good clean separate commits are much better than a squashed commit.

We currently don't have automatic benchmarking setup. It's something we should do.

**Pull request merged**